--- tacacs-F4.0.4.28-orig/tac_plus.service	1969-12-31 16:00:00.000000000 -0800
+++ tacacs-F4.0.4.28/tac_plus.service	2018-05-29 10:44:53.003596000 -0700
@@ -0,0 +1,22 @@
+[Unit]
+Description=TACACS+ IPv4 Daemon
+After=network.target
+After=crond.service
+ConditionPathExists=/etc/tac_plus.conf
+
+[Service]
+LimitCORE=16G
+StandardOutput=null
+
+ExecStartPre=/usr/sbin/tac_plus \
+  -C /etc/tac_plus.conf \
+  -P
+
+ExecStart=/usr/sbin/tac_plus \
+  -C /etc/tac_plus.conf \
+  -G
+
+Restart=always
+
+[Install]
+WantedBy=multi-user.target
